### DISCUSSION TITLE
Text Rendering Support

### DISCUSSION THREAD

**Discussion by cartesian-theatrics** [elalish/manifold#780] [Ideas]
Any thoughts on adding text support? We could consider doing a fairly clean ripoff the OpenSCAD solution based on freetype2.

https://github.com/openscad/openscad/blob/8951c9bf04f4bd807fedc16a8c10d2c4dd4db047/src/core/FreetypeRenderer.cc

**Comment by pca006132** [comment]
I don't think this is considered *core operation*, but maybe we can add it to the js/python bindings?

**Reply by hrgdavor** [reply]
For JS, you could add it to be available in the manifoldcad script execution enviroment, but it would not be advisable to bundle into the JS binding. When using manifold as JS library, I can import my own font handling library and load font form somewhere, then use manifold to do stuff with the shapes.

**Reply by pca006132** [reply]
yeah this makes sense, a separate package "manifoldjs-utils" seems a good idea.

**Reply by elalish** [reply]
Agreed, this seems like a good addition to ManifoldCAD - and could be made available as a package of sorts if we build `worker.ts` into our npm package. @cartesian-theatrics does any of that interest you?

**Reply by cartesian-theatrics** [reply]
Sounds good to me, except I kinda need it to be  implemented in c++ and consumable as part of some shared library. Ideally the util library would be available as a build target. Are you saying the implementations would be in python/js?

**Reply by elalish** [reply]
I think the main thing is we have no desire to maintain font libraries here. Anything that generates a polygon can easily enough be used in our library - there are lots of existing libraries to turn fonts into polygons. It's generally easiest for us to consume them in JS/Python since C++ dependencies are always a pain. Have you looked for a C++ font library yet?

**Reply by cartesian-theatrics** [reply]
That's fair. I looked a little while back for a font -> polygons library. Didn't find anything that did only that but I'm sure it's out there somewhere. I'll take a look.

**Reply by pca006132** [reply]
Maybe there are font -> SVG, and then you are use the SVG -> polygon code in jscad.

**Reply by cartesian-theatrics** [reply]
Since I'm not consuming from JS I'm thinking if I can't find an existing impl I'll probably write a c++ lib that converts freetype glyphs to polygons.

**Comment by cartesian-theatrics** [comment]
Created a small library to support this:

https://github.com/SovereignShop/text-to-polygon

And was able to integrate it with Manifold without too much trouble. Could create an PR if you're interested.

https://github.com/elalish/manifold/compare/master...SovereignShop:manifold:text-support?expand=1

It's very minimal at the moment. The only dependency is freetype, which I'm able to pull with FetchContent.

![Text](https://github.com/elalish/manifold/assets/26554374/14bcecb9-9ef3-48de-99ed-d9ed1315b943)

**Reply by elalish** [reply]
Thanks! I don't think we want this is in the core library, as fonts have a tendency to balloon into large binaries. However, I'd welcome a PR with links and information to make it easy to integrate by other projects. It might also be nice to have in ManifoldCAD.org - do you want to build this into an npm module, or does such a thing already exist?
